### ISSUE TITLE
[WINDOWS] XBOX Series X|S Controller input broken with Bluetooth (double button presses)

### ISSUE THREAD

**Issue by HVR88** [mmatyas/pegasus-frontend#799]
Input actions from buttons/sticks/pad on the newest XBOX Series X/S controllers (those with a Share button) are doubled when used with a Bluetooth connection.  

**UPDATE May 30 2020:** _This issue happens only with wireless connections, Bluetooth and MS Wireless Adapter. It doesn't occur when used with a directly-connected USB Cable._

Every time the D-Pad is clicked, the selection will move two times - happens to all 4 directions.
ABXY buttons same result - one press results in two actions
Both Analog Sticks - two movements for each press in every direction
Shoulder buttons (L1 and R1) same thing
LEFT Trigger button (L2) seems to react properly with ONE action only
RIGHT Trigger (R2) produces unexplained results - using GameOS while browsing a platform it should advance letters of the alphabet in order, but instead it immediately jumps to Z. It then has no effect if presses again, even after pressing the LEFT.

### System info
Windows 10, 202H, build 19042.804 

XBOX Series X Controller, Carbon Black, Connected via Bluetooth

### Pegasus version

Latest as of February 15 2021. Tried with default theme and GameOS.

**Comment by mmatyas** [comment]
Hi, this sounds similar to #783, do you experience similar behavior?

**Comment by HVR88** [comment]
I swear I looked, but didn't see the other issue. :)  Please feel free to merge them.

The only difference I can think of at the moment is I believe that I only have a single contoller device showing up in properties. Prior to the very latest update (mentioned in my OP) the controller showed multiple mis-indentified devices and didn't work at all in any software.

**Comment by HVR88** [comment]
Some additional info.

Using Antimicro in Windows, I can see that the controller shows up as two separate XInput Controllers. This app, like Pegasus is reacting twice (once per detected controller) for every button/action.

Other apps, like Cemu or Playnite (among others) don't see two and work as expected.

Only one device shows up in Device Manager (even after displaying hidden devices).

**Comment by mmatyas** [comment]
I've just updated the gamepad support in the latest release. Could you check if it improves the gamepad detection for you?

**Comment by HVR88** [comment]
The 15-3 version from May 30 2021 does not work with XBOX controllers at all in Windows.  No controller produces any action within Pegasus - keyboard remains functional.

I have gone back and verified that Alpha15 and 14-5 from February work properly.

The double button problem reported here seems to have been fixed by a Microsoft update since this issue was posted. I've verified that both XBOX One and newest XBOX Series X|S controllers function correctly and as expected in previous Pegasus releases where the problem was first reported.

**Comment by mmatyas** [comment]
Thanks for testing! So if I see correctly, things are worse than they were? That's unfortunate. Could you post the log files?
In general, the gamepad detection part is not written by me, so I can only hope too that the devs will improve it.

**Comment by HVR88** [comment]
In setting up to get log data, I've just undone all the testing results from earlier today. 

When I started testing earlier both my controllers were already paired with the MS Wireless adapter, so I tested that first. Then I removed the pairing and paired the controllers with bluetooth. Then I tried them with USB cables and back to Bluetooth a couple of times.  At that time all the results I posted were verifiable - no controller was working in the latest Pegasus build regardless of how it was connected. And they worked well/correctly in older builds with USB and MS Adapter.

This afternoon I had to re-pair with the MS adapter to get everything set up again and also of course restarted the system in the meantime. My results are now the following:

New build: Works with both controllers in all connection types!  YAY!
Old builds: XBOX One controller works same as always. XBOX Series controller only works properly with USB. MS Wireless adapter and Bluetooth BOTH do double actions.

LOGGING with only one controller connected via BLUETOOTH - XBOX Series X|S Controller

LOG line from new build:
	Line 28: 2021-05-30T16:17:39 [i] Gamepad: Connected device 0x0 (Xbox Series Controller)

LOG lines from Old builds:
	2021-05-30T16:21:28 [i] Gamepad: Connected device 0x0 (XInput Controller)
	2021-05-30T16:21:28 [i] Gamepad: Connected device 0x1 (Xbox Series Controller)

New build properly identifies the controller only once and as the right model. Old build identifies it twice, which is what causes the double actions.

I also tested using GamePad Tool (for making SDL maps) and it shows the same issue - it detects TWO controllers for the Series model. Interestingly it doesn't label either as the correct model "Series"

**Comment by HVR88** [comment]
Two buttons on Series X|S controllers still don't work with the newest SDL.  L2 and R2 (analog triggers) both produce undefined results.  In initial bug report, L2 worked correctly, now it doesn't.  R2 in GameOS still immediately advances to Z and then has the appearance of being stuck.

**Comment by mmatyas** [comment]
Ok, good to hear it's fixed at least partially! Technically the triggers are axes, with a pressure strength that you could use in eg. games. Whether SDL treats one as an axis or button depends on the gamepad mapping. Have you made any layout changes in Pegasus? In that case there should be an `sdl_controllers.txt` file in the config dir; perhaps removing it might fix the issue. If there isn't, or you haven't made yet, you could try so and set the L2/R2 buttons in Pegasus' Settings menu, hopefully this time it will work. You could also try creating the mapping with the Gamepad Tool and add its output to the `sdl_controllers.txt` manually.

**Comment by HVR88** [comment]
I definitely haven't made any mapping changes, but I'll take a look at creating one for the Series controller to see if I can work around the issue.

Can I also get you any debug output to show any difference at the Pegasus level between the XBOX One and Series controllers for these axis?

**Comment by HVR88** [comment]
Trying to change config inside Pegasus doesn't work properly with either XBOX controller.  With Series X, pressing the R2 causes L2 to light up in green and it stays green.

With XBOX One, it doesn't do the above, but using any AXIS input causes the on-onscreen corresponding control to light up in green and stay green - as if it's continuously receiving input.  

In all cases, regardless of which control is selected for change, nothing is ever updated on-screen to indicate a different mapping, so there's no way to tell what's going on.  For example, if I select the A button and instead press the B button, nothing changes on the screen.

With regards to setting an Axis as a button, I don't see anything inside the config file that indicates that's possible unless I was to simply map trigger as another EXISTING button (duplicating it).  In the SDL tool there everything is B0-whatever for device buttons and A0-whatever for device AXIS, with DPAD being H0.#-# for hat buttons.

**Comment by HVR88** [comment]
Maybe I should start a different but report for the Trigger buttons and close this one.

**Comment by mmatyas** [comment]
> Maybe I should start a different bug report for the Trigger buttons and close this one.

Sure, it seems the original duplicate devices issue got fixed. Meanwhile I'll check if there's an update for the SDL mappings too.

**Comment by HVR88** [comment]
Haven't forgotten, just been tied up - I'll create new issue soon about axis triggers.
